src/relentless/model/potential/bond.py

```diff
@@ -0,0 +1,425 @@
+import numpy
+
+from relentless import math
+
+from . import potential
+
+
+class BondParameters(potential.Parameters):
+    """Parameters for a bond potential."""
+
+    pass
+
+
+class BondPotential(potential.Potential):
+    r"""Abstract base class for a bond potential."""
+
+    pass
+
+
+class Harmonic(BondPotential):
+    r"""Harmonic bond potential.
+
+    .. math::
+
+        u(r) = \frac{k}{2} (r - r_0)^2
+
+    where :math:`r` is the distance between two bonded particles. The parameters
+    for each type are:
+
+    +-------------+--------------------------------------------------+-----------+
+    | Parameter   | Description                                      | Initial   |
+    +=============+==================================================+===========+
+    | ``k``       | potential constant                               |           |
+    +-------------+--------------------------------------------------+-----------+
+    | ``r0``      | rest length.                                     |           |
+    +-------------+--------------------------------------------------+-----------+
+
+    Parameters
+    ----------
+    type : str
+        Type.
+    name : str
+        Unique name of the potential. Defaults to ``__u[id]``, where ``id`` is the
+        unique integer ID of the potential.
+
+    Attributes
+    ----------
+    coeff : :class:`BondParameters`
+        Parameters of the potential for each type.
+
+    Examples
+    --------
+    Harmonic Bond::
+
+        >>> u = relentless.potential.bond.Harmonic("A")
+        >>> u.coeff["A"].update({'k': 1.0, 'r0': 1.0})
+
+    """
+
+    def __init__(self, types, name=None):
+        super().__init__(keys=types, params=("k", "r0"), name=name)
+
+    def energy(self, type, r):
+        """Evaluate bond energy."""
+        k = self.coeff[type]["k"]
+        r0 = self.coeff[type]["r0"]
+        return k / 2 * (r - r0) ** 2
+
+    def force(self, type, r):
+        """Evaluate bond force."""
+        k = self.coeff[type]["k"]
+        r0 = self.coeff[type]["r0"]
+        return -k * (r - r0)
+
+    def derivative(self, type, var, r):
+        r"""Evaluate bond derivative with respect to a variable."""
+        k = self.coeff[type]["k"]
+        r0 = self.coeff[type]["r0"]
+        if var == "k":
+            return (r - r0) ** 2 / 2
+        elif var == "r0":
+            return -k * (r - r0)
+        else:
+            raise ValueError("Unknown parameter")
+
+
+class FENE(BondPotential):
+    r"""The Finite Extensible Nonlinear Elastic (FENE) bond interaction.
+
+    .. math::
+
+        u(r) = -\frac{k r_0^2}{2} \ln \left(1 -
+            \left(\frac{r - r_0}{r_0}\right)^2\right)
```

Comment:
This should be written like this:
```suggestion
        u(r) = -\frac{k r_0^2}{2} \ln \left[1 -
            \left(\frac{r}{r_0}\right)^2\right]
```
I will check if the equations are right below or not